model/github_runner.rb

```diff
@@ -6,6 +6,7 @@ class GithubRunner < Sequel::Model
   one_to_one :strand, key: :id
   many_to_one :installation, key: :installation_id, class: :GithubInstallation
   one_to_one :vm, key: :id, primary_key: :vm_id
+  one_to_one :active_billing_record, class: :BillingRecord, key: :resource_id do |ds| ds.active end
```

Comment:
This is in the realm of terrible ideas, what if we create one billing record per project per month, instead of per github_runner. We need to change active_billing_record to return the record that is created in this month and also instead of creating new billing record for every github runner, they would just update the existing one by adding their duration to it.

Primary benefit would be generating single billing record showing the total duration in the invoice (well not single but one per runner size).

Reply:
I truly love the concept of having fewer billing records. When I aggregated all line items by description from lengthy invoices, @ozgune suggested aggregating them on a daily basis instead of all at once.
This means we could potentially have one billing record per project per day rather than per month.
However, I'm uncertain about one aspect: our granularity would decrease from minutes to days/months. I wonder if customers might want to see their usage broken down by the minute in the future?

Reply:
We use billing records for auditing as well. However, that's not a problem in this situation because we already have billing records for the underlying virtual machine and IPv4 address.

Reply:
I am all for moving in the direction of collapse of billing records. Sure, we can't get away from everything for system telemetry or being a good netizen for resolving abuse reports, but the residue of really short jobs (on Heroku, this was `heroku run`) tends to snowball into a sometimes ungratifyingly tricky data management problem.

Reply:
I said this is in the realm of terrible ideas, because we are "abusing" the billing records. If we do this end time will have a different meaning for github runners than the rest of resources, but I think that is OK.

These are good inputs for next iteration of billing, which should handle such needs more gracefully rather than requiring us to give second meaning to the fields. I think as we build new services, we will find new use cases that needs to be addressed. Another one is usage based billing instead of duration based one (network egress for example). I'm waiting a bit more before addressing such needs in more foundational way in the billing subsystem. Until then, I think what we discussed here is OK.

About aggregating daily vs monthly; I'm fine with either way. What is the usual practice in the industry?

Reply:
I am not sure. There is obviously an amplification of 30x for daily, but really the big win is from going from perhaps 100,000 records per month per customer (for certain value of customer) to thirty-one, or one, record(s).

We could go maximum on the aggregation to a monthly roll-up to keep the control plane database as small as possible, but in doing so, we probably increase our pressure in finding a place to put fine-grained historical entries to check the bill.

Reply:
Okay I think we have a case here where we need the fine grained data, so I'm going to retract my suggestion that we move towards aggregation before we have some kind of data warehousing mechanism...this does set up a bit of an tension against us, where we will have to find a solution before the tables get too large, and the tables getting large is correlated to github action churning.

We should chart the size growth, perhaps monthly, or something like that.

Reply:
If the customer creates 1000 runners daily, we generate monthly billing records as follows:
- Current main: 60k in the customer project
- New GitHub runner rates without aggregation: 60k in the internal project, 30k in the customer project.
- New GitHub runner rate with daily aggregation: 60k in the internal project, 30 in the customer project.

Here are the corrected sentences:

I agree that we should continue saving fine-grained data when we have a data warehouse.

I prefer daily aggregation over monthly for now. Its benefits outweigh its costs. It only adds 29 records but provides more insight to the customer. Telling a customer they used our service for 300,000 minutes this month is not very informative.

Our internal projects still generates thousands of billing records. We might introduce a new flag for projects and refrain from creating billing records for our internal projects until we have a data warehouse solution. We can use the deleted_record table for auditing.